Source/Classes/PINRemoteImageDownloadQueue.m

```diff
@@ -0,0 +1,145 @@
+//
+//  PINRemoteImageDownloadQueue.m
+//  PINRemoteImage
+//
+//  Created by Garrett Moon on 3/1/17.
+//  Copyright © 2017 Pinterest. All rights reserved.
+//
+
+#import "PINRemoteImageDownloadQueue.h"
+
+#import "PINURLSessionManager.h"
+#import "PINRemoteLock.h"
+
+@interface PINRemoteImageDownloadQueue ()
+{
+    NSUInteger _runningOperationCount;
+    PINRemoteLock *_lock;
+    
+    NSMutableArray <NSURLSessionDataTask *> *_highPriorityQueuedOperations;
+    NSMutableArray <NSURLSessionDataTask *> *_defaultPriorityQueuedOperations;
+    NSMutableArray <NSURLSessionDataTask *> *_lowPriorityQueuedOperations;
+}
+
+@end
+
+@implementation PINRemoteImageDownloadQueue
+
+@synthesize maximumNumberOfOperations = _maximumNumberOfOperations;
+
+- (PINRemoteImageDownloadQueue *)initWithMaximumNumberOfOperations:(NSUInteger)maximumNumberOfOperations
+{
+    if (self = [super init]) {
+        _maximumNumberOfOperations = maximumNumberOfOperations;
+        
+        _lock = [[PINRemoteLock alloc] initWithName:@"PINRemoteImageDownloadQueue Lock"];
+        _highPriorityQueuedOperations = [[NSMutableArray alloc] init];
+        _defaultPriorityQueuedOperations = [[NSMutableArray alloc] init];
+        _lowPriorityQueuedOperations = [[NSMutableArray alloc] init];
+    }
+    return self;
+}
+
+- (NSUInteger)maximumNumberOfOperations
+{
+    [self lock];
+        NSUInteger maximumNumberOfOperations = _maximumNumberOfOperations;
```

Comment:
For these scalar types, there's been some really surprisingly good results using std::atomic in ASLayoutElementStyle.  It would be the first such usage in the framework, and so is your call.  The upside is that complexity is low (it would be a net removal of lines of code), and help the framework fulfill a hardcore dedication to minimized overhead for the app using it.

Some code examples & profiles here: https://github.com/facebook/AsyncDisplayKit/pull/2413 . The profiles don't capture how much savings it has in blocking though, only runtime benefits.  The real advantage is that it causes the thread scheduler to not have to stop executing the thread and then need to re-start it later:


![image](https://cloud.githubusercontent.com/assets/565251/23680824/7b4fc826-0341-11e7-8095-a2736d9eb9dd.png)
![]